pkg/yurttunnel/pki/pki.go

```diff
@@ -0,0 +1,192 @@
+)
+
+// GenTGenTLSConfigUseCertMgrAndCertPool generates a TLS configuration
+// using the given certificate manager and x509 CertPool
+func GenTLSConfigUseCertMgrAndCertPool(
+	m certificate.Manager,
+	root *x509.CertPool) (*tls.Config, error) {
+	tlsConfig := &tls.Config{
+		// Can't use SSLv3 because of POODLE and BEAST
+		// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+		// Can't use TLSv1.1 because of RC4 cipher usage
+		MinVersion: tls.VersionTLS12,
+		RootCAs:    root,
+		ClientAuth: tls.RequireAnyClientCert,
+	}
+
+	tlsConfig.GetClientCertificate =
+		func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			cert := m.Current()
+			if cert == nil {
+				return &tls.Certificate{Certificate: nil}, nil
+			}
+			return cert, nil
+		}
+	tlsConfig.GetCertificate =
+		func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			cert := m.Current()
+			if cert == nil {
+				return &tls.Certificate{Certificate: nil}, nil
+			}
+			return cert, nil
+		}
+
+	return tlsConfig, nil
+}
+
+// GenRootCertPool generates a x509 CertPool based on the given kubeconfig,
+// if the kubeConfig is empty, it will creates the CertPool using the CA file
+func GenRootCertPool(kubeConfig, caFile string) (*x509.CertPool, error) {
+	if kubeConfig != "" {
+		// kubeconfig is given, generate the clientset based on it
+		if _, err := os.Stat(kubeConfig); os.IsNotExist(err) {
+			return nil, err
+		}
+
+		// load the root ca from the given kubeconfig file
+		config, err := clientcmd.LoadFromFile(kubeConfig)
+		if err != nil || config == nil {
+			return nil, fmt.Errorf("failed to load the kubeconfig file(%s), %v",
+				kubeConfig, err)
+		}
+
+		if len(config.CurrentContext) == 0 {
+			return nil, fmt.Errorf("'current context' is not set in %s",
+				kubeConfig)
+		}
+
+		ctx, ok := config.Contexts[config.CurrentContext]
+		if !ok || ctx == nil {
+			return nil, fmt.Errorf("'current context(%s)' is not found in %s",
+				config.CurrentContext, kubeConfig)
+		}
+
+		cluster, ok := config.Clusters[ctx.Cluster]
+		if !ok || cluster == nil {
+			return nil, fmt.Errorf("'cluster(%s)' is not found in %s",
+				ctx.Cluster, kubeConfig)
+		}
+
+		if len(cluster.CertificateAuthorityData) == 0 {
+			return nil, fmt.Errorf("'certificate authority data of the cluster(%s) is not set in %s",
+				ctx.Cluster, kubeConfig)
+		}
+
+		rootCertPool := x509.NewCertPool()
+		rootCertPool.AppendCertsFromPEM(cluster.CertificateAuthorityData)
+		return rootCertPool, nil
+	}
+
+	// kubeConfig is missing, generate the cluster root ca based on the given ca file
+	if caFile == "" {
+		return nil, errors.New("either a kubeconfig or a CA file must be given")
+	}
+	_, err := os.Stat(caFile)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("loading CA from %s fail: %s", caFile, err)
+	}
+	caPEM, err := ioutil.ReadFile(caFile)
+	if err != nil {
+		return nil, err
+	}
+
+	rootCertPool := x509.NewCertPool()
+	rootCertPool.AppendCertsFromPEM(caPEM)
+	return rootCertPool, nil
```

Comment:
would you use GenCertPoolUseCA func to refactor the code snippet?

Reply:
fixed